File: lib/wcc.js
"use strict";

const ws = require("ws");

/**
 * Throws an error if condition is false.
 * @private
 * @param { boolean } condition
 * @param { string } message
 */
function assert(condition, message) {
  if (!condition) throw new Error("WCC: " + message);
}

/**
 * Log to console with prefix
 * @private
 * @param {...any} args
 */
function wccLog(...args) {
  console.log("WCC:", ...args);
}

/**
 * Log to console with prefix
 * @private
 * @param {...any} args
 */
function wccWarn(...args) {
  console.warn("WCC:", ...args);
}

/**
 * Log to console with prefix
 * @param {...any} args
 */
function wccError(...args) {
  console.trace("WCC:", ...args);
}

//------------------------------------------------------------------------------
// TYPEDEFS
//------------------------------------------------------------------------------
/** @typedef {{ user: string, pass: string }} UserPass */
/** @typedef { "message" | "data" | "other" } ProtocolMessageType */
/** @typedef {{ name: string, content: string }} Payload */
/** @typedef {{ version: number, type: ProtocolMessageType, from: string, to: string[], payload: unknown }} ProtocolMessageUnknown */
/** @typedef {{ version: number, type: "message", from: string, to: string[], payload: Payload }} ProtocolMessage */

//------------------------------------------------------------------------------
// CLIENT
//------------------------------------------------------------------------------
class WebsocketCollabClient {
  //--------------------------------------------------------------------------
  // MEMBERS
  //--------------------------------------------------------------------------
  /**
   * Version of the protocol used for the communication from client to client.
   * @constant
   * @readonly
   * @type { number }
   */
  PROTOCOL_VERSION = 1;

  /**
   * If the client is currently connected to the server.
   * @readonly
   * @type { boolean }
   */
  connected = false;

  /**
   * ID of the joined channel.
   * @readonly
   * @type { string }
   */
  channel_id = "";

  /**
   * Url of the joined server.
   * @readonly
   * @type { URL? }
   */
  server_url = null;

  /**
   * @private
   * @type { string }
   */
  user = "";

  /**
   * @private
   * @type { ws? }
   */
  socket = null;

  /**
   * @private
   * @type { number }
   */
  retry_count = 0;

  /**
   * @private
   * @constant
   * @readonly
   * @type { number }
   */
  MAX_RETRIES = 5;

  //--------------------------------------------------------------------------
  // PRIVATE METHODS
  //--------------------------------------------------------------------------
  /**
   * @private
   * @param { ws.ErrorEvent } ev
   */
  handleWsError(ev) {
    wccError("Collab websocket error:", ev.type, ev.message);
    wccWarn(
      "Lost connection to Collab Websocket due to error, trying reconnect..."
    );
    this.reconnect();
  }

  /**
   * @private
   * @param { ws.CloseEvent } ev
   */
  handleWsClose(ev) {
    if (!ev.wasClean) {
      wccWarn("Lost connection to Collab Websocket, trying reconnect...");
      this.reconnect();
      return;
    }
  }

  /**
   * @private
   * @param { ws.MessageEvent } ev
   */
  handleWsMessage(ev) {
    let data = ev.data.toString("utf8");

    /** @type { unknown | undefined } */
    let obj;
    try {
      obj = JSON.parse(data);
    } catch {
      return;
    }

    if (!this.isWellFormedProtocolMessage(obj)) {
      return;
    }

    /** @type { ProtocolMessageUnknown } */
    let proto_msg = obj;

    if (!this.isWellFormedMessagePayload(proto_msg.payload)) {
      return;
    }

    /** @type { ProtocolMessage } */
    let msg = proto_msg;

    if (this.onAllMessages) {
      this.onAllMessages(msg);
    }

    if (!(msg.to.includes("all") || msg.to.includes(this.user))) {
      // Should not be received by user.
      return;
    }

    if (msg.from == this.user) {
      // Is from user
      return;
    }

    switch (msg.type) {
      case "message": {
        if (!this.onTextMessage) return;
        this.onTextMessage(msg.payload.name, msg.payload.content, msg);
        return;
      }
      case "data": {
        if (!this.onDataMessage) return;
        this.onDataMessage(msg.payload.name, msg.payload.data, msg);
        return;
      }
    }
  }

  /**
   * @private
   * @param { URL } url
   * @returns { Promise<void> }
   */
  connectWebsocket(url) {
    return new Promise((resolve, reject) => {
      let resolved = false;

      this.socket = new ws(url);

      this.socket.onerror = (err) => {
        wccError(
          "Error while trying to connect to server:",
          err.type,
          err.message
        );
        if (resolved) return;
        resolved = true;
        reject(err.message);
      };

      this.socket.onopen = (ev) => {
        this.socket.ping = this.socket.pong();
        this.socket.onerror = this.handleWsError.bind(this);
        this.socket.onmessage = this.handleWsMessage.bind(this);
        this.socket.onclose = this.handleWsClose.bind(this);
        this.connected = true;
        this.retry_count = 0;
        if (resolved) return;
        resolved = true;
        resolve();
      };
    });
  }

  /**
   * @private
   * @param { unknown } obj
   * @returns { boolean }
   */
  isWellFormedProtocolMessage(obj) {
    if (obj.version == undefined || typeof obj.version != "number")
      return false;
    if (obj.type == undefined || typeof obj.type != "string") return false;
    if (obj.from == undefined || typeof obj.from != "string") return false;
    if (obj.to == undefined || typeof obj.to != "object") return false;
    if (obj.to.length == undefined) return false;
    if (obj.payload == undefined) return false;
    return true;
  }

  /**
   * @private
   * @param { unknown } payload
   * @returns { boolean }
   */
  isWellFormedMessagePayload(payload) {
    if (payload.name == undefined || typeof payload.name != "string")
      return false;
    if (payload.content == undefined || typeof payload.content != "string")
      return false;
    return true;
  }

  /**
   * @private
   */
  clearServerData() {
    this.user = "";
    this.channel_id = "";
    this.server_url = null;
    this.connected = false;
  }

  /**
   * @private
   */
  reconnect() {
    if (this.retry_count >= this.MAX_RETRIES) {
      wccError(
        "Max retries reached. Could not reconnect to the Websocket Collab server."
      );
      this.clearServerData();
      return;
    }

    let delay_sec = Math.pow(2, this.retry_count) * 1000;
    wccLog(`Reconnecting in ${delay_sec / 1000} seconds...`);
    this.retry_count++;

    if (this.socket) {
      this.socket.removeAllListeners();
      this.socket.close();
      this.socket = null;
    }

    setTimeout(() => {
      this.connectWebsocket(this.server_url)
        .then(() => {
          wccLog("Reconnected successfully.");
        })
        .catch((err) => {
          wccError("Reconnection attempt failed:", err);
          this.reconnect(); // Retry again
        });
    }, delay_sec);
  }

  //--------------------------------------------------------------------------
  // PUBLIC METHODS
  //--------------------------------------------------------------------------
  /**
   * Connects to the server and joins the selected channel.
   *
   * Rejects if fails to establish connection to server.
   * @method
   * @function
   * @param { string | URL } url Url to the collab server.
   * @param { string } channel_id ID of the channel to join.
   * @param { UserPass } auth User and Password.
   * @returns { Promise<void> }
   */
  connect(url, channel_id, auth) {
    assert(url, "argument url cannot be undefined.");
    assert(channel_id, "argument channel_id cannot be undefined.");
    assert(auth, "argument auth cannot be undefined.");
    assert(auth.user, "argument auth.user cannot be undefined.");
    assert(auth.pass, "argument auth.pass cannot be undefined.");

    assert(
      url instanceof URL || typeof url == "string",
      "argument url must be of type string or URL."
    );
    assert(
      typeof channel_id == "string",
      "argument channel_id must be of type string."
    );
    assert(
      typeof auth.user == "string",
      "argument auth.user must be of type string."
    );
    assert(
      typeof auth.pass == "string",
      "argument auth.user must be of type string."
    );

    let uri;
    if (typeof url != "string") uri = url;
    else uri = new URL(url);

    uri.username = encodeURI(auth.user);
    uri.password = encodeURI(auth.pass);
    uri.pathname = `/${channel_id}`;

    this.user = auth.user;
    this.server_url = uri;
    this.channel_id = channel_id;

    return this.connectWebsocket(uri);
  }

  /**
   * Closes the connection to the server.
   * @method
   * @function
   */
  disconnect() {
    wccLog("Disconnected from Websocket Collab Server.");
    if (this.socket.readyState != ws.CLOSED) {
      this.socket.close();
    }
  }

  /**
   * Send a message to users of the channel, with the assumption that they sould
   * respond to said message.
   * @method
   * @function
   * @param { string } sender_name Name of the sender, for ex: "Hilda" or "Meteora"
   * @param { string } content Content of the text message.
   * @param { string[] } to Users to send the message to. If contains `"all"`,
   * will send message to all participants.
   */
  sendText(sender_name, content, to = ["all"]) {
    assert(
      sender_name != undefined && typeof sender_name == "string",
      "argument sender_name must be a string."
    );
    assert(
      content != undefined && typeof content == "string",
      "argument content must be a string."
    );
    assert(
      to != undefined && typeof to == "object",
      "argument to must be of type string array."
    );

    assert(to.length != undefined, "argument to must be of type string array.");
    for (let name of to) {
      assert(
        name != undefined && typeof name == "string",
        'in argument to: "' + name + '" must be of type string.'
      );
    }

    if (!(this.socket != null && this.socket.readyState == ws.OPEN)) {
      wccWarn("Cannot send to collab, connection to server is closed.");
      return;
    }

    /** @type { ProtocolMessage } */
    let obj = {
      version: 1,
      type: "message",
      from: this.user,
      to: to,
      payload: {
        name: sender_name,
        content: content,
      },
    };

    this.socket.send(JSON.stringify(obj));
  }

  /**
   * Send data to other participants, with the assumption that they may or may not
   * respond to receiving data.
   * @method
   * @function
   * @param { string } label Identifier of the sent data.
   * @param { string } data Sent data.
   * @param { string } to Users to send the message to. If contains `"all"`,
   * will send message to all participants.
   */
  sendData(label, data, to = ["all"]) {
    assert(
      label != undefined && typeof label == "string",
      "argument label must be a string."
    );
    assert(
      data != undefined && typeof data == "string",
      "argument data must be a string."
    );
    assert(
      to != undefined && typeof to == "object",
      "argument to must be of type string array."
    );

    assert(to.length != undefined, "argument to must be of type string array.");
    for (let name of to) {
      assert(
        name != undefined && typeof name == "string",
        'in argument to: "' + name + '" must be of type string.'
      );
    }

    if (!(this.socket != null && this.socket.readyState == ws.OPEN)) {
      wccWarn("Cannot send to collab, connection to server is closed.");
      return;
    }

    /** @type { ProtocolMessage } */
    let obj = {
      version: 1,
      type: "data",
      from: this.user,
      to: to,
      payload: {
        name: label,
        content: data,
      },
    };

    this.socket.send(JSON.stringify(obj));
  }

  /**
   * Called when a protocol message is received from the server.
   *
   * Is called before `onTextMessage` and `onDataMessage`, so they might also
   * get called for the same message.
   *
   * Only use if you want to handle specific cases.
   *
   * Use `onMessage` instead if you are only interested with the text messages.
   * @member
   * @type { ((json: ProtocolMessage) => void)? }
   */
  onAllMessages = null;

  /**
   * Called when a text message is received from the server.
   *
   * The assumption is that you will respond to this message.
   * @method
   * @type { ((sender: string, content: string, raw_json: ProtocolMessage) => void)? }
   */
  onTextMessage = null;

  /**
   * Called when a data message is received from the server.
   *
   * The assumption is that you will not respond to this message.
   * @method
   * @type { ((label: string, data: string, raw_json: ProtocolMessage) => void)? }
   */
  onDataMessage = null;

  removeAllListeners() {
    this.onAllMessages = null;
    this.onTextMessage = null;
    this.onDataMessage = null;
  }
}

module.exports = WebsocketCollabClient;

File: lib/wcc.py
import urllib3.util
import json
import sys
import threading
import time

from typing import Any, Callable

import websockets
import websockets.exceptions
import websockets.sync
import websockets.sync.client

class Payload:
    name: str
    content: str

    def __init__(self,  name: str, content: str) -> None:
        self.name = name
        self.content = content

    def to_dict(self)-> dict:
        return {
            "name": self.name,
            "content": self.content
        }



class ProtocolMessageUnknown:
    version: int
    type: str
    sender: str
    to: list[str]
    payload: Any

    def __init__(self, version: int, type: str, sender: str, to: list[str], payload: Any ) -> None:
        self.version = version
        self.type = type
        self.sender = sender
        self.to = to
        self.payload = payload
    
    def to_dict(self)-> dict:
        return {
            "version": self.version,
            "type": self.type,
            "from": self.sender,
            "to": self.to,
            "payload": self.payload
        }



class ProtocolMessage:
    version: int
    type: str
    sender: str
    to: list[str]
    payload: Payload

    def __init__(self, version: int, type: str, sender: str, to: list[str], payload: Payload ) -> None:
        self.version = version
        self.type = type
        self.sender = sender
        self.to = to
        self.payload = payload
    
    def to_dict(self)-> dict:
        return {
            "version": self.version,
            "type": self.type,
            "from": self.sender,
            "to": self.to,
            "payload": self.payload.to_dict()
        }



class WebsocketCollabClient:

    PROTOCOL_VERSION: int = 1
    """
    Version of the protocol used for the communication from client to client.
    """

    connected: bool = False
    """
    If the client is currently connected to the websocket server.
    """

    channel_id: str = ""
    """
    ID of the channel the client is currently connected to.
    """

    server_url: str = ""
    """
    URL of the server the client is currently connected to.
    """

    __retry_count: int = 0
    MAX_RETRIES: int = 5

    __user: str = ""
    __socket: websockets.sync.client.ClientConnection | None = None
    __thread: threading.Thread | None = None
    __should_close_thread: bool = False

    __listeners_all_msg: list[Callable[[ProtocolMessage], None]] = []
    __listeners_text_msg: list[Callable[[ProtocolMessage], None]] = []
    __listeners_data_msg: list[Callable[[ProtocolMessage], None]] = []


    def __wcc_log(*args):
        print("WCC:", *args)
    

    def __wcc_err(*args):
        print("WCC:", *args, file=sys.stderr)


    def __connect_websocket(self, url: str)-> None:
        self.__socket = websockets.sync.client.connect(url)
        self.connected = True
    

    def __clear_server_data(self)-> None:
        self.__user = ""
        self.channel_id = ""
        self.server_url = ""
        self.connected = False


    def __is_well_formed_protocol_message(self, obj: dict)-> bool:
        if "version" not in obj or type(obj["version"]).__name__ != "int":
            return False
        if "type" not in obj or type(obj["type"]).__name__ != "str":
            return False
        if "from" not in obj or type(obj["from"]).__name__ != "str":
            return False
        if "to" not in obj or type(obj["to"]).__name__ != "list":
            return False
        if "payload" not in obj:
            return False
        return True


    def __is_well_formed_payload(self, obj: dict)-> bool:
        if "name" not in obj or type(obj["name"]).__name__ != "str":
            return False
        if "content" not in obj or type(obj["content"]).__name__ != "str":
            return False
        return True


    def __call_listeners(self, listeners: Callable, arg: Any)-> None:
        for listener in listeners:
            listener(arg)


    def __listener(self)-> None:
        while not self.__should_close_thread:
            received: str
            try:
                received = str(self.__socket.recv())
            except (websockets.exceptions.ConnectionClosedError, OSError):
                self.connected = False
                self.__reconnect()
                continue
            except Exception:
                continue

            obj: dict
            try:
                obj = json.loads(received)
            except:
                continue
            
            if not self.__is_well_formed_protocol_message(obj):
                continue

            proto_msg: ProtocolMessageUnknown = ProtocolMessageUnknown(
                version=obj["version"],
                type=obj["type"],
                sender=obj["from"],
                to=obj["to"],
                payload=obj["payload"]
            )

            if not self.__is_well_formed_payload(proto_msg.payload):
                continue
            
            payload = Payload(obj["payload"]["name"], obj["payload"]["content"])

            msg: ProtocolMessage = ProtocolMessage(
                version=proto_msg.version,
                type=proto_msg.type,
                sender=proto_msg.sender,
                to=proto_msg.to,
                payload=payload
            )

            self.__call_listeners(self.__listeners_all_msg, msg)

            if not ("all" in msg.to or self.__user in msg.to):
                continue

            if msg.sender == self.__user:
                continue

            match msg.type:
                case "message":
                    self.__call_listeners(self.__listeners_text_msg, msg)
                    continue
                case "data":
                    self.__call_listeners(self.__listeners_data_msg, msg)
                    continue
                
    
    def __start_listener_thread(self)-> None:
        self.__thread = threading.Thread(target=self.__listener, args=[])
        self.__thread.start()
    

    def __reconnect(self) -> None:
        if self.__retry_count >= self.MAX_RETRIES:
            self.__wcc_err("Max retries reached. Could not reconnect to the Websocket Collab server.")
            self.__clear_server_data()
            return

        delay_sec = 2 ** self.__retry_count
        self.__retry_count += 1
        self.__wcc_log(f"Reconnecting in {delay_sec} seconds...")

        time.sleep(delay_sec)

        try:
            self.__connect_websocket(self.server_url)
            self.__retry_count = 0
            self.__wcc_log("Reconnected successfully.")
        except websockets.exceptions.InvalidURI:
            self.__wcc_err("Invalid URI for reconnection attempt.")
            self.__clear_server_data()
        except Exception as err:
            self.__wcc_err(f"Reconnection attempt failed: {err}")
            self.__reconnect()


    def connect(self, url: str | urllib3.util.Url, channel_id: str, user: str, password: str):
        """
        Connects to the server and joins the selected channel.
        """
        assert url != None
        assert channel_id != None
        assert user != None
        assert password != None

        assert type(url).__name__ == "str" or type(url).__name__ == "Url"
        assert type(channel_id).__name__ == "str"
        assert type(user).__name__ == "str"
        assert type(password).__name__ == "str"

        uri: urllib3.util.Url
        if (type(url).__name__ == "str"):
            uri = urllib3.util.parse_url(url)
        else:
            uri = url
        
        reconstructed_uri: str = ""

        if uri.scheme is not None:
            reconstructed_uri += uri.scheme + "://"
        reconstructed_uri += f"{user}:{password}" + "@"
        if uri.host is not None:
            reconstructed_uri += uri.host
        if uri.port is not None:
            reconstructed_uri += ":" + str(uri.port)
        reconstructed_uri += f"/{channel_id}"

        self.__user = user
        self.channel_id = channel_id
        self.server_url = reconstructed_uri
        
        self.__connect_websocket(reconstructed_uri)
        self.__start_listener_thread()
    

    def disconnect(self)-> None:
        """
        Closes the connection to the server.
        """
        self.__clear_server_data()
        if self.__socket == None: return
        self.__should_close_thread = True
        self.__thread.join()
        self.__socket.close()
    

    def __send(self, msg_type: str, payload: dict, to: list[str] | None = None)-> None:
        assert msg_type != None
        assert payload != None

        if to == None: to = ["all"]
        
        assert type(msg_type).__name__ == "str"
        assert type(payload).__name__ == "dict"
        assert type(to).__name__ == "list"

        assert self.__socket != None

        obj = {
            "version": 1,
            "type": msg_type,
            "from": self.__user,
            "to": to,
            "payload": payload,
        }

        try:
            self.__socket.send(json.dumps(obj))
        except:
            self.__wcc_err("Failed to send message to Websocket Collab server.")
    

    def send_text(self, sender_name: str, content: str, to: list[str] | None = None)-> None:
        """
        Send a message to users of the channel, with the assumption that they
        should respond to said message.
        """
        self.__send("message", {
            "name": sender_name,
            "content": content
        }, to)


    def send_data(self, data_name: str, data: str, to: list[str] | None = None)-> None:
        """
        Send data to other participants, with the assumption that they will not
        respond to receiving the data.
        """
        self.__send("data", {
            "name": data_name,
            "content": data
        }, to)
    

    def on_all_messages(self, listener: Callable[[ProtocolMessage], None])-> None:
        """
        Called when a protocol message is received from the server.

        Is called before `on_text_message` and `on_data_message`, so they might
        also get called for the exact same message.

        Only use if you want to handle specific cases.

        Use `on_text_message` instead if you are only interested with the text
        messages.
        """
        self.__listeners_all_msg.append(listener)
    

    def on_text_message(self, listener: Callable[[ProtocolMessage], None])-> None:
        """
        Called when a text message is received from the server.

        The assumption is that you will respond to this message.
        """
        self.__listeners_text_msg.append(listener)
    

    def on_data_message(self, listener: Callable[[ProtocolMessage], None])-> None:
        """
        Called when a data message is received from the server.

        The assumption is that you will not respond to this message.
        """
        self.__listeners_data_msg.append(listener)


    def remove_all_listeners(self)-> None:
        """
        Remove all listeners added via `on_#_message` functions.
        """
        self.__listeners_all_msg.clear()
        self.__listeners_text_msg.clear()
        self.__listeners_data_msg.clear()
